tls/s2n_client_hello.c

```diff
@@ -258,11 +258,37 @@ static int s2n_parse_client_hello(struct s2n_connection *conn)
 }
 
 bool s2n_is_tls_12_self_downgrade_required(struct s2n_connection *conn) {
-    /* RSA PSS is required for TLS 1.3 connections. So if there's a possibility that an RSA Certificate could be picked
-     * by a client connection, then downgrade connection to TLS 1.2 if our libcrypto doesn't support RSA PSS. */
-    return ((conn->mode == S2N_SERVER)
-              && conn->config->is_rsa_cert_configured
-              && !s2n_is_rsa_pss_signing_supported());
+    if ((conn->mode == S2N_SERVER)
+          && (conn->config != NULL)
+          && conn->config->is_rsa_cert_configured
+          && !s2n_is_tls13_fully_supported()) {
+        /* RSA PSS is required for TLS 1.3 connections if RSA is used. So if we are a server that doesn't support RSA PSS,
+         * and there's a possibility that an RSA Certificate could be picked by a client connection, then self-downgrade
+         * connection to TLS 1.2. */
+        return true;
+    }
+
+    s2n_cert_auth_type client_auth_status = S2N_CERT_AUTH_NONE;
+    s2n_connection_get_client_auth_type(conn, &client_auth_status);
+
+    if ((conn->mode == S2N_SERVER)
+        && (client_auth_status != S2N_CERT_AUTH_NONE)
+        && !s2n_is_tls13_fully_supported()) {
```

Comment:
Is this tested anywhere?

Reply:
Not that I'm aware of. I can create a backlog issue to test this case if you'd like.

Reply:
Opened https://github.com/aws/s2n-tls/issues/3049